designs/0013-profiling.md

```diff
@@ -0,0 +1,420 @@
+situation is was important to resolve this ambiguity, then the timing
+statements should be moved from inside the function to where the function
+is used:
+
+```
+model {
+  real x;
+  profile("myfunc - model") {
+    x = myfunc(alpha);
+  }
+  ...
+}
+...
+generated quantities {
+  real x;
+  profile("myfunc - generated quantities") {
+    x = myfunc(alpha);
+  }
+  ...
+}
+```
+
+## Note on parameters block
+
+Because only parameter declaration statements are allowed in the `parameters`
+block, `profile` cannot be used there. This means that the `lower`, `upper`,
+`multiply` and `offset` constraints/transforms can not be profiled.
+
+# The Stan Math implementation
+
+The `profile` functionality in Stan is implemented witho two functions, `profile_start`
+and `profile_stop` which handle intrumenting the code. A C++ `std::map` is to
+store the accumulated timing results of each profile by name and thread ID.
+
+The two functions `profile_start` and `profile_stop` are implemented with the
+signatures:
+
+```
+template <typename T>
+inline void profile_start(std::string name, profile_map& profiles);
+template <typename T>
+inline void profile_stop(std::string name, profile_map& profiles);
+```
+
+The first argument is the profile name, and the second argument is the map
+in which to store timing results.
+
+`profile_start` starts a timer for the given profile name and
+`profile_stop` stops the timer and saves the accumulated time in the profile.
+
+It is an error if `profile_start` is called with the same name twice before
+a `profile_stop` is called.
+
+If the template argument `T` is a `var`, then `profile_start` and
+`profile_stop` push varis onto the chaining autodiff stack. On the reverse
+pass, the vari produced by `profile_stop` starts a timer and the vari
+produced by `profile_start` stops it and records the results. In this way
+the forward and reverse pass calculations can both be timed.
+`profile_start` and `profile_stop` also record the number of varis on
+the chaining and non-chaining stacks.
+
+If the template argument `T` is `double`, then `profile_start` and
+`profile_stop` only record timing information. They do not push any
+varis onto the chaining autodiff stack or record information about the
+number of varis on the chaining and non-chaining autodiff stacks.
+
+A helper profile class is used to wrap `profile_start` and `profile_stop`
+with RAII semantics to make it easier to use.
+
+# Stanc3 - changes to the generated C++
+
+The C++ code generated from Stan models will be changed accordingly:
+
+- All generated C++ models representing Stan models will gain a private member
+`profiles__`. This member will represent the map of all profile information.
+
+- C++ calls to profile will be generated as:
+
+    ```cpp
+    {
+      profile<T__> profile("literal", const_cast<profile_map&>(profiles__));
+      ... statements
+    }    
+    ```
+    where X will be replaced by a unique number. The `const_cast` is required
```

Comment:
Where is `X`?

Reply:
Ah, we no longer require these to be unique as

```
{
  profile<var> profile__("name1", ...);
  {
   profile<var> profile__("name2", ...);
  }
}
```
is valid C++ and we cant have two profiles in the same level block. Will remove.